test/functional/Valhalla/src/org/openj9/test/lworld/NullRestrictedTypeOptTests.java

```diff
@@ -343,4 +343,43 @@ static private void evalTestEA8(int i) {
 			escape8 = p;
 		}
 	}
+
+   public static class TestStoreToNullRestrictedField {
+      public PrimPair nullRestrictedInstanceField;
+      public static PrimPair nullRestrictedStaticField;
+
+      public void replaceInstanceField(Object val) {
+         nullRestrictedInstanceField = (PrimPair) val;
+      }
+
+      public static void replaceStaticField(Object val) {
+         nullRestrictedStaticField = (PrimPair) val;
+      }
+   }
+
+   public static primitive class TestWithFieldStoreToNullRestrictedField {
+      public PrimPair nullRestrictedInstanceField;
+
+      public TestWithFieldStoreToNullRestrictedField(Object val) {
+         this.nullRestrictedInstanceField = (PrimPair) val;
+      }
+   }
+
+   @Test(expectedExceptions = java.lang.NullPointerException.class)
+   static public void testStoreNullValueToNullRestrictedInstanceField() throws Throwable {
+      new TestStoreToNullRestrictedField().replaceInstanceField(new PrimPair(1, 2));
+      new TestStoreToNullRestrictedField().replaceInstanceField(null);
+   }
+
+   @Test(expectedExceptions = java.lang.NullPointerException.class)
+   static public void testStoreNullValueToNullRestrictedStaticField() throws Throwable {
+      TestStoreToNullRestrictedField.replaceStaticField(new PrimPair(1, 2));
+      TestStoreToNullRestrictedField.replaceStaticField(null);
+   }
+
+   @Test(expectedExceptions = java.lang.NullPointerException.class)
+   static public void testWithFieldStoreToNullRestrictedField() throws Throwable {
+      new TestWithFieldStoreToNullRestrictedField(new PrimPair(1, 2));
+      new TestWithFieldStoreToNullRestrictedField(null);
+   }
```

Comment:
It might be a good idea to split each of these test methods in two to verify that the NullPointerExceptions are only thrown if a null reference is stored to the null-restricted field.

Reply:
> split each of these test methods in two ...

Could you clarify a bit more? I'm not sure I follow which two methods to split. For example, `TestStoreToNullRestrictedField().replaceInstanceField` is called twice because the first invocation is interpreted with count=1 and the first call to `TestStoreToNullRestrictedField().replaceInstanceField` doesn't store a NULL to the null restricted field.

Reply:
I was worried about the possibility that a bug could cause the first call to `replaceInstanceField` to throw a NullPointerException.  That would be wrong, but the test would still pass because the exception was thrown somewhere in the method.  One way to more closely narrow the range for the exception would be something like this:

```
   static public void testStoreNullValueToNullRestrictedInstanceField() throws Throwable {
      TestStoreToNullRestrictedField storeFieldObj = new TestStoreToNullRestrictedField();
      storeFieldObj.replaceInstanceField(new PrimPair(1, 2));
      try {
         storeFieldObj.replaceInstanceField(null);
      } catch (NullPointerException npe) {
         return;
      } finally {
         assertTrue(false);  // No exception or wrong kind of exception thrown
      }
   }
```

Anyway, the case that I was worried about probably isn't that likely or important - maybe it's not worth complicating the tests to handle it.

Reply:
Thank you the clarification! It's a valid concern. I'll explicitly test the exception from the second invocation as illustrated on the above.

Reply:
Addressed in 7725103